File: lr/src/questions/q_0543_diameter_of_binary_tree.rs
use std::{cell::RefCell, rc::Rc};

use crate::utils::TreeNode;

pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    let mut result = 0;
    diameter_of_binary_tree_helper(root, &mut result);
    result
}

fn diameter_of_binary_tree_helper(root: Option<Rc<RefCell<TreeNode>>>, result: &mut i32) -> i32 {
    if root.is_none() {
        return -1;
    }

    let left =
        diameter_of_binary_tree_helper(root.as_ref().unwrap().borrow_mut().left.take(), result);
    let right =
        diameter_of_binary_tree_helper(root.as_ref().unwrap().borrow_mut().right.take(), result);

    *result = i32::max(*result, 2 + left + right);

    1 + i32::max(left, right)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tree;

    #[test]
    fn test_diameter_of_binary_tree() {
        assert_eq!(diameter_of_binary_tree(tree!(1, 2)), 1);
        assert_eq!(diameter_of_binary_tree(tree!(1, 2, 3, 4, 5)), 3);
    }
}

File: lg/src/questions/q_0543_diameter-of-binary-tree.go
package questions

import "github.com/hi-rustin/lg/src/utils"

func diameterOfBinaryTree(root *utils.TreeNode) int {
	res := 0

	var dfs func(node *utils.TreeNode) int
	dfs = func(node *utils.TreeNode) int {
		if node == nil {
			return 0
		}
		left := dfs(node.Left)
		right := dfs(node.Right)
		res = max(res, left+right)
		return 1 + max(left, right)
	}

	dfs(root)
	return res
}
